### DISCUSSION TITLE
Attention Weights

### DISCUSSION THREAD

**Discussion by sincyvt** [pyg-team/pytorch_geometric#6049] [Q&A]
Is it possible to provide edge weights/ node weights as attention scores (α_ij between nodes i and j)directly, instead of calculatiing attention scores

**Comment by rusty1s** [comment]
In that case, you can leverage `GraphConv` and simply input an additional `edge_weight` vector to your layer.

**Reply by sincyvt** [reply]
in GraphConv, if edge_weight is provided, is it needed to overwrite forward function? 
can we expect GraphConv to output something else other than nodefeautre(NxF)

**Reply by rusty1s** [reply]
No, you don't need to overwrite anything, and it will return a node feature matrix of shape `NxF`.

**Reply by sincyvt** [reply]
during graph classification(batch mode), what will be the shape of node feature, edge index and edge weight. for example if batch size is 32, then node feature consists of XxF where X =  total number of nodes in 32 graphs, F = feature dim. edge index size 2xY where Y  = total number of edges in 32 graphs and edge weight = Y. Is this assumption right? How can we add edge feature matrix instead of edge weight?

**Reply by rusty1s** [reply]
Yes, your understanding is correct. You can also use multi-dimensional edge features (we call them `edge_attr` in PyG), but not all provided GNN layers support them.

**Reply by sincyvt** [reply]
but during batch training  the model differentiate between nodes using data.batch, which nodes belongs to which graph.
what about those edges? if number of nodes is Y = 400, how the model understand what all edges belongs to first graph, second and so on, same in the case of edge_weight/ edge attribute. What about their types?
thanks for your fastest reply

**Reply by rusty1s** [reply]
During message passing, such differentiation is not necessary as edges never point from one graph to another. If you also want to perform global pooling across edges, you can get a `batch` for the edges via `batch[edge_index[0]]`.

**Reply by sincyvt** [reply]
ok sir, thank you. 
In the case of knowledge graph, which is the better model to embed both relations (edges) and entities(nodes).

**Reply by rusty1s** [reply]
Between which models? :) I think it is always a good idea to start with something simple like `RGCNConv` and go from there.

**Reply by sincyvt** [reply]
Sir,  In my knowledge graph I have  node features as embeddings of node entities. Along with that weights are assigned for each node. Is it possible to incorporate both node weights as well as node features to GCN model? How it can be made possible.

**Reply by rusty1s** [reply]
I think there exists many options to integrate those:
* Simply concatenate your node weights to your node feature matrix
* Re-weight node features based on node weights
* Map node weights to edge weights, e.g., via `edge_weight = (node_weight[edge_index[0]] + node_weight[edge_index[1]]) / 2`

**Reply by sincyvt** [reply]
any sample code available?

**Reply by rusty1s** [reply]
No, this is a special use-case :)

* Concatenation: `x = torch.cat([data.x, data.node_weight.view(-1, 1)], dim=-1)`
* Re-weighting: `x = data.node_weight.view(-1, 1) * data.x`
* To edge weights: see above.

**Reply by sincyvt** [reply]
sir, 
during gcn training, some loss calculation steps are there.
 loss = criterion(predicted, actual)
if the task is question answering, during training the answer node should get higher score. Is there any way for training such tasks

**Reply by rusty1s** [reply]
Did you already look at recent literature for question answering?

**Reply by sincyvt** [reply]
If we are using GraphConv for a node embedding purpose, i.e., updating initial node embeddings by passing it through 1 or 2 GraphConv layers, how the training occurs. How to fix weights    W1, W2. Is it possible to obtain updated node embeddings without training by simply passing through some fixed number of layers?

**Reply by sincyvt** [reply]
how to learn weights for node/ edge embedding ? any sample code available?

**Reply by rusty1s** [reply]
Usually GNN weights need to be trained according to some supervised/self-supervised/unsupervised loss formulation, although an untrained GNN has already strong discriminating power. There exists a lot of examples in PyG on how to train GNNs (basically all examples in the `examples/` folder), e.g., `examples/gcn.py` (for supervised), `graph_sage_unsup.py` or `infomax_inductive.py` (for unsupervised). Ideally, you train your GNN end-to-end according to your given downstream-task.

**Reply by sincyvt** [reply]
is it possible to add more training data to gcn like models, once the model is already trained as like in model.fit() in keras

**Reply by rusty1s** [reply]
Sure, you can continue training it on more data if that is what you mean :)

**Reply by sincyvt** [reply]
is it the same function, our_model.fit()? any examples/ sample code/ link available?

**Reply by rusty1s** [reply]
Sorry, not really sure what you mean. Can you clarify? :(

**Reply by sincyvt** [reply]
Suppose l  trained and created my GCN model. Is it then possible to add more training data to the existing model instead of building it from the scratch. If yes, how it is, what will be the function for adding training data again. Any examples available?

**Reply by rusty1s** [reply]
I think you would train it the same way, put start from the trained model rather than an uninitialized one. You can load the trained model parameters by saving and loading your model via `torch.save` and `torch.load` for example.

**Reply by sincyvt** [reply]
Is it possible to fit validation data?

**Reply by rusty1s** [reply]
Sure, you can do that, but then you can no longer evaluate on that data. You also need to take into account that NNs are good at forgetting prior knowledge so if you simply fit on validation data, it will probably forget about knowledge from the training data. At best, fine-tune on validation+train data.

**Reply by sincyvt** [reply]
how to fine-tune on validation+train data? Also in the case of RGCN, is it possible to incorporate multi dimensional edge features? If yes how? what is edge_type in RGCN?

**Reply by rusty1s** [reply]
You would do something like
```python
train_val_mask = train_mask | val_mask
```
and continue training on this set. Does this make sense?

`RGCN` currently cannot take edge features into account (by definition).

**Reply by sincyvt** [reply]
which will be a better model that consider multi dimensional edge features into consideration(embeddings of edge labels are considered as feature here)

**Reply by rusty1s** [reply]
You can try `GENConv` or `TransformerConv`.

**Reply by sincyvt** [reply]
Similarly is it possible to incorporate 'edge_attr' parameter in GCNConv or EdgeConv. 
or which model best consider both edge weight and edge_attr

**Reply by sincyvt** [reply]
Any sample code for using class GCN form models

https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GCN.html#torch_geometric.nn.models.GCN

**Reply by sincyvt** [reply]
#My model

class GCNNew(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels =128, num_layers =2, out_channels = 128):
        super(GCNNew, self).__init__()
        torch.manual_seed(12345)
        print(in_channels)
        self.conv1 = GCN(num_node_features, in_channels,num_layers)
        self.conv2 = GCN(in_channels, out_channels, num_layers)
        self.lin = Linear(in_channels, 2)

    def forward(self, x, edge_index, batch):
        x = self.conv1(x, edge_index, batch)
        x = x.relu()
        x = self.conv2(x, edge_index, batch)
        x = x.relu()
        x = F.dropout(x, p=0.5, training=self.training)
        y = self.lin(x)
        return x,y

model_newg1 = GCNNew(in_channels = 128)
print(model_newg1)



def train():
    model_newg1.train()
    optimizer.zero_grad()
    out, pred = model_newg1(x, data.edge_index,  data.batch)
    F.nll_loss(pred, data.y).backward()
    optimizer.step()


A sample code for basic GCN model as per  "https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GCN.html#torch_geometric.nn.models.GCN"  that I made was given 

the error here is "TypeError: BasicGNN.forward() takes 3 positional arguments but 4 were given"
What modification should I have to make to get rid from this error
Also how to include edge_weight and edge_attr. When I am trying to include them same error is showing with difference in number of parameters. Can you help me to correct the mistake that I am making

**Reply by sincyvt** [reply]
def forward(self, x, edge_index, edge_weight, edge_attr):
        x = self.conv1(self, x, edge_index,  edge_weight, edge_attr)
        x = x.relu()
        x = self.conv2(self, x, edge_index, edge_weight, edge_attr)
        x = x.relu()
        x = F.dropout(x, p=0.5, training=self.training)
        y = self.lin(x)
        return x,y


out, pred = model_newg1(x, data.edge_index,  data.edge_attr, data.edge_weig)


If I changed my forward function like this, the error is 
TypeError: BasicGNN.forward() takes 3 positional arguments but 6 were given

What change should I have to make?

**Reply by rusty1s** [reply]
I don't think we have support for a GNN layer that incorporates both `edge_weight` and `edge_attr`. What people usually do in this case is just concatenate the features: `edge_attr= torch.cat([edge_attr, edge_weight.view(-1, 1)], dim=-1)`.

**Reply by sincyvt** [reply]
GCN supported by edge_attr or edge_weight? I think it is edge_weight

**Reply by sincyvt** [reply]
then how to include this "edge_attr= torch.cat([edge_attr, edge_weight.view(-1, 1)], dim=-1)" since edge_attr of dimension (no:of edges, no:of edge_features), is not accepted by GCN. GCN accepts only edge_weigh of dimension (no: of edges, 1). Is this correct?

**Reply by sincyvt** [reply]
Thankyou so much for answering my questions. While testing GENCONV, each time the result varies. What may be the expected reason

**Reply by rusty1s** [reply]
Yeah, GCN only supports single edge weights, so you could use any GNN layer that supports multi-dimensional edge features. Can you clarify what you mean with varying results? Do you fix the seed? Other than that, a small variance is expected since reductions are non-deterministic since the order of floating point operations is undetermined.

**Reply by sincyvt** [reply]
> Thankyou so much for answering my questions. While testing GENCONV, each time the result varies. What may be the expected reason

Solved. thankyou.

**Reply by sincyvt** [reply]
GCNConv, GENConv, TransformerConv these all are convolutional layers. Will you please explain the need of a convolutional layer? How it affect a model? Can we define different models by simply changing this convolutional layers only

**Reply by rusty1s** [reply]
Yes, these are different GNN layers/message passing layers proposed in literature, which different strengths and properties, e.g., making use of attention when aggregating neighborhood information.

**Comment by sincyvt** [comment]
"import torch_geometric.data" produces the following error 
OSError: /home/hp2/anaconda3/envs/myphd/lib/python3.7/site-packages/torch_sparse/_diag_cuda.so: undefined symbol: _ZN3c106detail19maybe_wrap_dim_slowEllb
Any suggestions to solve it

**Reply by rusty1s** [reply]
This suggests that your `torch-sparse` version is incompatible with your installed PyTorch version.

**Reply by sincyvt** [reply]
How to resolve it?

**Reply by rusty1s** [reply]
Depends on how you installed `torch-sparse`. I suggest to run `pip uninstall torch-sparse` repeatedly, and then try to re-install it by ensuring you install the right package according to `torch.__version__` and `torch.version.cuda`.

**Comment by sincyvt** [comment]
Sir, 
Is there any Language Model applications like BERT to be combined with GNN. i.e, whether it is any GNN models directly available for natural language processing(NLP)

**Reply by akihironitta** [reply]
I'd suggest creating another discussion to get more visibility from the community as your question doesn't seem to be relevant to this discussion. (I'd recommend RelBench where they combine tabular data (that can potentially have textual columns) with GNNs. https://github.com/snap-stanford/relbench/blob/45b9d782d44deafeb08d517bfc5ac4e72b4a0eec/examples/gnn.py)
